### DISCUSSION TITLE
Explicit cast should suppress warning for nullable types

### DISCUSSION THREAD

**Discussion by clairernovotny** [dotnet/csharplang#2221] [Language Ideas] [answered]
This is with the version in Preview 2.

Consider the following

```csharp

void Foo(string bar)
{
}

void Frob(string? bbb)
{
    Foo((string)bbb); // Still generates warnings: uncessesary cast, CS8600, CS8604
}
```

There are many times when a nullable type may need to go into a place that's marked non-nullable. I would expect an explicit cast to work without warnings. It currently does not.

Here's my reasoning: an explicit cast is the developer saying "do this and throw if it breaks". It's also cleaner and inline with the language expressing intent. The alternative, of using pragma's around that line to suppress, is far more invasive to the code and makes it harder to read.

Here's the pragmas:
```csharp
void Frob(string? bbb)
{
#pragma warning disable CS8600 // Converting null literal or possible null value to non-nullable type.
#pragma warning disable CS8604 // Possible null reference argument.
    Foo(bbb); 
#pragma warning restore CS8604 // Possible null reference argument.
#pragma warning restore CS8600 // Converting null literal or possible null value to non-nullable type.
}
```

I think that's much worse.

**Comment by BillWagner** [comment]
Why not use this construct to indicate "do this and throw if it breaks"? 

```csharp
void Foo(string bar)
{
}

void Frob(string? bbb)
{
    Foo(bbb ?? throw new ArgumentNullException()); 
}
```

**Comment by clairernovotny** [comment]
It's still more code and seems to be more complex to read. That pattern also changes the logic, putting the checking on me. A cast would merely suppress the warning and let the existing logic/checking handle things whatever it might be.  Null may actually be okay.

**Comment by DynConcepts** [comment]
I despise "case to clear warning" types with the C# style of casts. They are virtually impossible to reliably find in a code base [this is not a problem with C++ type cases (e.g. static_cast<T>(...) ]

**Comment by cston** [comment, marked as answer]
The post-fix `!` operator will suppress nullability warnings and the compiler will treat the expression as non-null.
```C#
    Foo(bbb!); // no warnings
```

**Comment by jaredpar** [comment]
Moving to csharplang as this is a language design question. The language team has already made a decision here and the compiler team is just following through on the implementation.

**Comment by clairernovotny** [comment]
While I don't care for the `!` post-fix, it could just be new....  I think it'd help a lot of a Roslyn quick-fix offered to do that in addition to the pragmas.

![image](https://user-images.githubusercontent.com/1427284/52733022-a7968880-2f8f-11e9-9fc3-5cfe9615b9a2.png)

This seems harder to read :/

![image](https://user-images.githubusercontent.com/1427284/52733140-f7754f80-2f8f-11e9-9682-55730f4723c3.png)

**Comment by HaloFour** [comment]
@onovotny 

Did you try putting the post-fix operator after the method call?  It looks like the compiler is warning on assigning the `object?` result from `GetValueFromLocalSettings<T>(name)` to the variable `object value`, which sounds right to me.

**Comment by clairernovotny** [comment]
@HaloFour there are two places where the post-fix operator is required and adding them does make the warning go away.

**Comment by clairernovotny** [comment]
I've filed an issue to track a quick-fix for the `!` post-fix operator: https://github.com/dotnet/roslyn/issues/33360
